### ISSUE TITLE
Issue Running Ludwig AutoML on Modal.com Cloud Computing Service

### ISSUE THREAD

**Issue by degschta** [ludwig-ai/ludwig#3544]
Hello everyone,

I am currently attempting to use Ludwig's AutoML method on the Modal.com cloud computing platform, but I keep running into an error. I've tried to search for a solution in the existing issues and documentation, but couldn't find a similar problem.

Environment:

    Ludwig version: 0.8.1.post1
    modal client version: 0.51.3143
    OS: Ubuntu 20.04

Steps to Reproduce:

    Install modal client in your local environment
    Use modal's `@stub.function()` annotation to run Ludwig's `auto_train()` method on modal's compute cluster.

Below you can find the full code example to reproduce the error:

```
from modal import Image, Stub, Mount

image = Image.debian_slim().pip_install(
    # scraping pkgs
    "scikit-learn==1.3.0",
    "InstructorEmbedding",
    "unstructured",
    "llama-cpp-python",
    "torch",
    "sentence-transformers",
    "ludwig[full]",
    "pandas")

stub = Stub(
    name="example-langchain-qanda",
    image=image,
)
@stub.function(gpu='a100', mounts=[Mount.from_local_dir("data", remote_path="/root/remove_boilerplate")], timeout=7100)
def ludwig_auto_train():
    from ludwig.automl import auto_train
    from ludwig.utils.dataset_utils import get_repeatable_train_val_test_split
    import pandas as pd

    artistery_df = pd.read_json("/root/remove_boilerplate/searchresult.anno.jsonl", lines=True)
    artistery_df = get_repeatable_train_val_test_split(artistery_df, 'is_artist_website', random_seed=42)

    auto_train_results = auto_train(
        dataset=artistery_df,
        target='is_artist_website',
        time_limit_s=7000,
        tune_for_memory=False,
        user_config={'preprocessing': {'split': {'column': 'split', 'type': 'fixed'}}},
    )

@stub.local_entrypoint()
def main():
    print("the result is", ludwig_auto_train.call())
```

Expected Behavior:
I expected Ludwig's AutoML to train the model without any issues.

Actual Behavior:
Instead, I received the following error:

```
== Status ==
Current time: 2023-08-25 12:15:29 (running for 00:12:17.65)
Using AsyncHyperBand: num_stopped=2
Bracket: Iter 1800.000: None | Iter 360.000: None | Iter 72.000: 0.8377622365951538
Logical resource usage: 0/128 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:A100)
Current best trial: 3f4a1e14 with metric_score=0.94405597448349 and parameters={'trainer.learning_rate': 2.2103375806114724e-05, 'trainer.batch_size': 64, 'combiner.num_fc_layers': 1, 'combiner.output_size': 128, 'combiner.dropout': 0.012855425737772442}
Result logdir: /root/hyperopt
Number of trials: 5/5 (5 TERMINATED)
+----------------+------------+--------------------+--------------------+------------------------+------------------------+----------------------+------------------------+--------+------------------+----------------+
| Trial name     | status     | loc                |   combiner.dropout |   combiner.num_fc_laye |   combiner.output_size |   trainer.batch_size |   trainer.learning_rat |   iter |   total time (s) |   metric_score |
|                |            |                    |                    |                     rs |                        |                      |                      e |        |                  |                |
|----------------+------------+--------------------+--------------------+------------------------+------------------------+----------------------+------------------------+--------+------------------+----------------|
| trial_c49a6168 | TERMINATED | 172.19.5.202:12580 |          0.0093055 |                      3 |                    256 |                 1024 |            3.10601e-05 |     11 |         174.323  |       0.898601 |
| trial_bfc4c69b | TERMINATED | 172.19.5.202:12852 |          0.086701  |                      2 |                    256 |                 1024 |            0.000337902 |     11 |         162.256  |       0.881119 |
| trial_3f4a1e14 | TERMINATED | 172.19.5.202:13075 |          0.0128554 |                      1 |                    128 |                   64 |            2.21034e-05 |     11 |         175.502  |       0.944056 |
| trial_2602beac | TERMINATED | 172.19.5.202:13300 |          0.0393212 |                      1 |                    128 |                 1024 |            0.000705436 |      5 |          75.9226 |       0.72028  |
| trial_b482c408 | TERMINATED | 172.19.5.202:13504 |          0.0782563 |                      3 |                    128 |                 1024 |            0.000517778 |      5 |          75.7609 |       0.825175 |
+----------------+------------+--------------------+--------------------+------------------------+------------------------+----------------------+------------------------+--------+------------------+----------------+

2023-08-25 11:54:22,603	INFO tune.py:945 -- Total run time: 732.09 seconds (732.06 seconds for the tuning loop).
Traceback (most recent call last):
  File "/pkg/modal/_container_entrypoint.py", line 351, in handle_input_exception
    yield
  File "/pkg/modal/_container_entrypoint.py", line 437, in run_inputs
    res = imp_fun.fun(*args, **kwargs)
  File "/root/test_ludwig.py", line 70, in ludwig_auto_train
    auto_train_results = auto_train(
  File "/usr/local/lib/python3.10/site-packages/ludwig/automl/automl.py", line 157, in auto_train
    return train_with_config(dataset, config, output_directory=output_directory, random_seed=random_seed, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/ludwig/automl/automl.py", line 278, in train_with_config
    hyperopt_results = _train(
  File "/usr/local/lib/python3.10/site-packages/ludwig/automl/automl.py", line 399, in _train
    hyperopt_results = hyperopt(
  File "/usr/local/lib/python3.10/site-packages/ludwig/hyperopt/run.py", line 349, in hyperopt
    hyperopt_results = hyperopt_executor.execute(
  File "/usr/local/lib/python3.10/site-packages/ludwig/hyperopt/execution.py", line 931, in execute
    self._evaluate_best_model(
  File "/usr/local/lib/python3.10/site-packages/ludwig/hyperopt/execution.py", line 426, in _evaluate_best_model
    eval_stats, _, _ = best_model.evaluate(
  File "/usr/local/lib/python3.10/site-packages/ludwig/api.py", line 1019, in evaluate
    eval_stats, predictions = predictor.batch_evaluation(
  File "/usr/local/lib/python3.10/site-packages/ludwig/models/predictor.py", line 254, in batch_evaluation
    outputs = self._predict_on_inputs(inputs)
  File "/usr/local/lib/python3.10/site-packages/ludwig/models/predictor.py", line 324, in _predict_on_inputs
    return self.dist_model(inputs)
  File "/usr/local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/ludwig/models/ecd.py", line 135, in forward
    encoder_outputs = self.encode(inputs)
  File "/usr/local/lib/python3.10/site-packages/ludwig/models/ecd.py", line 75, in encode
    encoder_output = encoder(input_values)
  File "/usr/local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/ludwig/features/text_feature.py", line 207, in forward
    encoder_output = self.encoder_obj(inputs_exp, mask=inputs_mask)
  File "/usr/local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/ludwig/encoders/text_encoders.py", line 671, in forward
    transformer_outputs = self.transformer.module(
  File "/usr/local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/transformers/models/bert/modeling_bert.py", line 1015, in forward
    embedding_output = self.embeddings(
  File "/usr/local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/transformers/models/bert/modeling_bert.py", line 232, in forward
    inputs_embeds = self.word_embeddings(input_ids)
  File "/usr/local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/torch/nn/modules/sparse.py", line 162, in forward
    return F.embedding(
  File "/usr/local/lib/python3.10/site-packages/torch/nn/functional.py", line 2210, in embedding
    return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argument index in method wrapper_CUDA__index_select)
2023-08-25 11:54:31,161	WARNING tune.py:184 -- Stop signal received (e.g. via SIGINT/Ctrl+C), ending Ray Tune run. This will try to checkpoint the experiment state one last time. Press CTRL+C (or send SIGINT/SIGKILL/SIGTERM) to skip. 
Runner terminated, in-progress inputs will be re-scheduled`
```

Additional Information:
    The training seems to terminate with a `RuntimeError` only after all 5 trials finish successfully. This is particularly strange because the error indicates that the tensors used for training are scattered over different devices. If that was really the case, I would expect this to become an issue during AutoML training, not after termination.

Has anyone else faced a similar issue? I would appreciate any help or guidance to resolve this.

Thank you!

**Comment by tgaddair** [comment]
Hey @degschta, thanks for reporting this issue! It looks like somehow the model weights are ending up on the CPU during evaluation. Let me take a look and if we can repro the issue on our side.

**Comment by tgaddair** [comment]
Looks like this might have been a bug introduced in #3362, @jeffkinnison is going to take a look today at getting in a fix for a 0.8.2 release!

The general idea of the fix should be to add back a call to `distrbuted.to_device()` in the `Predictor`.

**Comment by jeffkinnison** [comment]
Hi @degschta! Just reaching out to let you know I've tracked down the source of the problem and am working on a fix right now.

**Comment by tgaddair** [comment]
Fixed by #3554.
